### PR TITLE
Feature/instance include dataset

### PR DESCRIPTION
**Pull request by CarlHembrough**
### What

Refactored to call the recipe API to determine how many instances to create, and also to have a dataset ID for each instance.

- Added job.service class to hold the business logic for adding / updating jobs
- Added recipe.API client
- Add url.builder to hold hostnames and url structure
- removed some redundant parameters / methods
- renamed some things to make them more consistent
- removed unused vendor dependencies

### How to review

- Review code
- Check that it still works

### Who can review

Anyone

### CODE REVIEW

File: models/models.go [hunk 1]
```diff
@@ -15,8 +15,8 @@ type JobResults struct {
 
 // Job for importing datasets
 type Job struct {
-	JobID         string          `bson:"id,omitempty"             json:"id,omitempty"`
-	Recipe        string          `bson:"recipe,omitempty"         json:"recipe,omitempty"`
+	ID            string          `bson:"id,omitempty"             json:"id,omitempty"`
+	RecipeURL     string          `bson:"recipe,omitempty"         json:"recipe,omitempty"`
```
Comment (Matt-Guest) [inline models/models.go]: Should this create the following JSON structure { "links": { "recipe": { "id":"123-456", "href":"http://........." }}}?

File: models/models.go [hunk 2]
```diff
@@ -41,6 +42,29 @@ func (job *Job) Validate() error {
...
 
+//this should probably be replaced with an import of
+//github.com/ONSdigital/dp-code-list-api/{pkg when unstubbed}
+type Recipe struct {
+	ID    string `json:"id"`
+	Alias string `json:"alias"`
+	//Format          string     `json:"format"`
+	//InputFiles      []file     `json:"files"`
+	OutputInstances []RecipeInstance `json:"output_instances"`
+	//Links           *links     `json:"links"`
```
Comment (Matt-Guest) [inline models/models.go]: There commented out lines here

File: models/models.go [hunk 3]
```diff
@@ -41,6 +42,29 @@ func (job *Job) Validate() error {
...
+}
+
+type RecipeInstance struct {
+	DatasetID string `json:"dataset_id"`
+	//Editions  []string   `json:"editions"`
```
Comment (Matt-Guest) [inline models/models.go]: Here as well

File: recipe/api.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+func (api *API) GetRecipe(url string) (*models.Recipe, error) {
+
+	logData := log.Data{"URL": url}
+
+	jsonResult, httpCode, err := api.get(url, nil)
+	logData["httpCode"] = httpCode
+	logData["jsonResult"] = jsonResult
+
+	if err == nil && httpCode != http.StatusOK && httpCode != http.StatusCreated {
```
Comment (Matt-Guest) [inline recipe/api.go]: Why check for a 201 code (created)?

File: job/service.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+func NewService(dataStore datastore.DataStorer, jobQueue JobQueue, datasetAPI DatasetAPI, recipeAPI RecipeAPI, urlBuilder *url.Builder) *Service {
+	return &Service{
+		dataStore:  dataStore,
+		jobQueue:   jobQueue,
+		datasetAPI: datasetAPI,
+		recipeAPI:  recipeAPI,
+		urlBuilder: urlBuilder,
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (CarlHembrough) [reply job/service.go]: Added 👍

File: models/models.go (hunk 2 above)
Comment (CarlHembrough) [reply models/models.go]: Done

File: models/models.go (hunk 3 above)
Comment (CarlHembrough) [reply models/models.go]: Done

File: recipe/api.go (hunk 4 above)
Comment (CarlHembrough) [reply recipe/api.go]: Copy and paste crimes 😞

File: models/models.go (hunk 1 above)
Comment (CarlHembrough) [reply models/models.go]: Good point, it seems like a reasonable suggestion - will discuss this further before changing.

Comment (CarlHembrough) [reply models/models.go]: Item added to tech debt log to be picked up when we implement the recipe API.

**Comment by Matt-Guest** [comment]
👍

**Pull request merged**